### DISCUSSION TITLE
Should Customizer objects use (JSONPath, replacement-value) or patch style?

### DISCUSSION THREAD

**Discussion by MikeSpreitzer** [kubestellar/kubestellar#453] [General]
In the current customization API, a Customizer has a series of [Replacements](https://github.com/kcp-dev/edge-mc/blob/v0.2.2/pkg/apis/edge/v1alpha1/customize.go#L79-L84); each one is a pair of ([JSONPath](https://goessner.net/articles/JsonPath/index.html) pointer to location in object, JSON replacement value).  JSONPath has the advantage that it is a well established in general and in Kubernetes (you can use it in kubectl output, for example) and expressive way of selecting part of an object.  This (pointer, replacement) style of customization is also widely used.  However, JSONPath is not _all_ that widely established.  I went looking for an implementation in go that supports modification of a data structure that was produced by `encoding/json.Unmarshal(str, map[string]interface{})` and came up empty.  The closest I found was [one](github.com/tidwall/gjson) that supports modifications but (a) the input and output are strings rather than data structures and (b) the pointer syntax is only a close approximation of a subset of JSONPath.  The current implemenation of customization uses that.  Someone could write a better implementation for our purposes.  Note that https://goessner.net/articles/JsonPath/index.html incorporates "Expressions of the underlying scripting language" and "Filter expressions".  These could be invitations to quite a lot of work.  But in the use case here I think we do not need support for all possible expressions.  Since the author of the Customizer can look at the object before customization, literal indices will suffice; one could go a little further for convenience, making a tradeoff between implementation and convenience.

A possible alternative approach would be to use patching a la [kubectl patch](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/).  This is also well established in the Kubernetes milieu.  Note that `kubectl patch` supports three different kinds of patch: [JSON Patch](https://tools.ietf.org/html/rfc6902), [JSON merge patch](https://tools.ietf.org/html/rfc7386), and kube's own unique "strategic merge patch".  We would not necessarily have to support all three; we could pick just one.  Which one?  If I had to pick just one, it would be strategic merge patch because it is the one _I_ use most often.  Also for this approach, I do not know a good re-usable implementation.  You can see what the apiserver does at https://github.com/kubernetes/apiserver/blob/be7d2eb4732cd3718f71ddd5a9eb3c74ab04d40e/pkg/endpoints/handlers/patch.go#L366-L413 for the first two and at https://github.com/kubernetes/apiserver/blob/be7d2eb4732cd3718f71ddd5a9eb3c74ab04d40e/pkg/endpoints/handlers/patch.go#L423-L447 for the last one.  That server-side code is thoroughly intertwined with apiserver stuff.

**Comment by MikeSpreitzer** [comment]
FYI, following are some go-based implementations of JSONPath that I found.

- https://pkg.go.dev/github.com/yalp/jsonpath - says to use https://github.com/PaesslerAG/jsonpath instead.
- https://github.com/PaesslerAG/jsonpath - includes expressions but does not support updates.
- https://github.com/oliveagle/jsonpath - "bleeding edge", does not support updates.
- https://github.com/spyzhov/ajson - works with parsed data, supports updates.
- https://github.com/limiu82214/gojmapr - about marshal/unmarshal directed by JSONPath in field tags.
- https://github.com/limiu82214/gojpath - supports reading but not updates.
- https://github.com/ohler55/ojg - works on parsed data (not sure if the usual types), supports updates.

**Comment by clubanderson** [comment]
it appears, based on your notes, that ohler55 and spyzhov look like a good start.  Are either of these communities that would be responsive to issues/PRs?
